pretyping/reductionops.ml

```diff
@@ -89,11 +104,22 @@ module ReductionBehaviour = struct
     | UnfoldWhen x -> UnfoldWhen (more_args_when k x)
     | UnfoldWhenNoMatch x -> UnfoldWhenNoMatch (more_args_when k x)
 
+(* We need to have a fast way to know the set of all constants that
+  have the NeverUnfold flag.  Therefore, the table has a distinct subpart
+  that is this set. *)
   let table =
-    Summary.ref (GlobRef.Map.empty : t GlobRef.Map.t) ~name:"reductionbehaviour"
+    Summary.ref ((GlobRef.Set.empty, GlobRef.Map.empty) :
+                   GlobRef.Set.t * t GlobRef.Map.t) ~name:"reductionbehaviour"
 
   let load _ (_,(_,(r, b))) =
-    table := GlobRef.Map.add r b !table
+    table := ((match b with
+                NeverUnfold ->
+                GlobRef.Set.add r (fst !table)
+                | _ -> GlobRef.Set.remove r (fst !table)),
+                       (match b with
+                         NeverUnfold ->
+                         GlobRef.Map.remove r (snd !table)
+                        |_ -> GlobRef.Map.add r b (snd !table)))
```

Comment:
```suggestion
    table := (match b with
                | NeverUnfold -> GlobRef.Set.add r (fst !table), GlobRef.Map.remove r (snd !table)
                | _ -> GlobRef.Set.remove r (fst !table), GlobRef.Map.add r b (snd !table))
```

I rephrase the code so that the invariant @ppedrot was talking about is more explicit